spacy/kb/candidate.py

```diff
@@ -0,0 +1,118 @@
+import abc
+from typing import List, Union, Callable
+
+
+class Candidate(abc.ABC):
+    """A `Candidate` object refers to a textual mention (`alias`) that may or may not be resolved
+    to a specific `entity_id` from a Knowledge Base. This will be used as input for the entity_id linking
+    algorithm which will disambiguate the various candidates to the correct one.
+    Each candidate (alias, entity_id) pair is assigned a certain prior probability.
+
+    DOCS: https://spacy.io/api/kb/#candidate-init
+    """
+
+    def __init__(
+        self,
+        mention: str,
+        entity_id: int,
+        entity_name: str,
+        entity_vector: List[float],
+        prior_prob: float,
+    ):
+        """Initializes properties of `Candidate` instance.
+        mention (str): Mention text for this candidate.
+        entity_id (int): Unique entity ID.
+        entity_name (str): Entity name.
+        entity_vector (List[float]): Entity embedding.
+        prior_prob (float): Prior probability of entity for this mention - i.e. the probability that, independent of
+            the context, this mention resolves to this entity_id in the corpus used to build the knowledge base. In
+            cases in which this isn't always possible (e.g.: the corpus to analyse contains mentions that the KB corpus
+            doesn't) it might be better to eschew this information and always supply the same value.
+        """
+        self._mention = mention
+        self._entity_id = entity_id
+        self._entity_name = entity_name
+        self._entity_vector = entity_vector
+        self._prior_prob = prior_prob
+
+    @property
+    def entity(self) -> int:
+        """RETURNS (int): Unique entity ID."""
+        return self._entity_id
+
+    @property
+    def entity_(self) -> str:
+        """RETURNS (int): Entity name."""
+        return self._entity_name
+
+    @property
+    def mention(self) -> str:
+        """RETURNS (str): Mention."""
+        return self._mention
+
+    @property
+    def entity_vector(self) -> List[float]:
+        """RETURNS (List[float]): Entity vector."""
+        return self._entity_vector
+
+    @property
+    def prior_prob(self) -> float:
+        """RETURNS (List[float]): Entity vector."""
+        return self._prior_prob
+
+
+class InMemoryCandidate(Candidate):
+    """Candidate for InMemoryLookupKB."""
+
+    def __init__(
+        self,
+        retrieve_string_from_hash: Callable[[int], str],
+        entity_hash: int,
+        entity_freq: int,
+        entity_vector: List[float],
+        alias_hash: int,
```

Comment:
`alias` and `mention` are used interchangeabily here, right? I suggest we stick with one?

Reply:
That's true. I'd go with mention then, since `alias` doesn't apply anymore. Agreed?

Reply:
I don't really get why we need the additional `alias methods on this class, maybe I'm missing something.

Reply:
So far I kept them for backwards compatibility. It's true that they aren't used in `EntityLinker` anywhere (otherwise we'd have to have this in `Candidate` as well anyway). Without `alias` and `alias_` won't have the possibility to retrieve any information about the alias in an instance of `InMemoryCandidate`, which would be annoying, I think?

Reply:
but the properties `_alias` and `mention` are the same, right? So maybe we should just add one method `mention_hash`?

Reply:
Ah, that's what you mean. Yes, that makes sense!

Reply:
Do we want to keep the `alias` terminology in `InMemoryLookupKB`?

Reply:
I'd say yes, since it still makes sense there.

Reply:
Updates to variable and property naming in https://github.com/explosion/spaCy/pull/12341/commits/46fe069f8731c3d591963fe4a1e3e1c1a4b1eef9 and https://github.com/explosion/spaCy/pull/12341/commits/94e57d0ed5fe1981cf5ac2b54964d0e3f14533a4.